cephfs: Implement VolumeGroupSnapshot API's

Madhu-1 opened the pull request:
This PR implements the CreateVolumeGroupSnapshot and DeleteVolumeGroupSnapshot API for cephfs.

pending items are captured here https://github.com/ceph/ceph-csi/issues/4435, will send separate PR's for it.

Madhu-1 committed:
cephfs: add volumegroup service capability

Add GROUP_CONTROLLER_SERVICE capabilities to
the GetPluginCapabilities of the cephFS
plugin.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: advertise group snapshot capability

Advertise VOLUME_GROUP_SNAPSHOT capability
from the cephfs driver.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: add validateCreateVolumeGroupSnapshotRequest

added validateCreateVolumeGroupSnapshotRequest
to validate the CreateVolumeGroupSnapshotRequest
request and ensure that all the requirement
options are set. if not, reject the RPC request.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: initialize VolumeGroupJournal

initialize VolumeGroupJournal which is
required for volumegroup rados communication

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: register group controller

register the group controller service
for the cephfs.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: add locks for volumegroup

Adding a lock for the volumegroup so
that we can take care of serializing
the same requests to ensure same requests
are not served in parallel.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: add error for quiesce operation

added ErrInProgress to indicate the
the quiesce operation is in progress.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: add helper for quiesce api

added helper function which calls
the go-ceph API for the quiesce
operations.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: add helper for group options

volumegroup.go holders all the helpers
to extra the group details from the request
and also to extra group details from the
groupID.

This also provide helpers to reserve group
for the request Name and also an undo function
incase if somethings goes wrong and we need to
cleanup the reserved omap entries.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: implement CreateVolumeGroupSnapshot RPC

implemented CreateVolumeGroupSnapshot RPC which
does below operations

* Basic request validation
* Reserve the UUID for the group name
* Quiesce the filesystem for all the subvolumes
from the input volumeId's
* Take the snapshot for all the input volumeId's
* Add the mapping between volumeId's and snapshot
Id's in omap
* Release the quiesce for the filesystem for
all the subvolumes from the input volumeId's

Undo all the operations if anything fails.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: implement DeleteVolumeGroupSnapshot RPC

implemented DeleteVolumeGroupSnapshot RPC which
does below operations

* Basic request validation
* Get the snapshotId's and volumeId's
mapping reserved for the UUID
* Delete snapshot and remove its mapping
from the omap
* Repeat above steps until all the mapping
are removed
* Remove the reserved uuid from the omap
* Reset the filesystem quiesce, This might be
required as cephfs doesnt provide any options to
remove the quiesce, if we get any request with same
ID again we can reuse the quiesce API for same set-id
* Return success if the received error is
Pool not found or key not found.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
util: add UnimplementedGroupControllerServer

adding UnimplementedGroupControllerServer to
the DefaultControllerServer struct to avoid
build errors when some non mandatory RPC's
are not implemented.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
Test results from local testing with hacks

```bash
[🎩︎]mrajanna@li-2cfbef4c-22d9-11b2-a85c-a3e4a93c405f cephfs $]kubectl get volumegroupsnapshot
NAME                                                     READYTOUSE   VOLUMEGROUPSNAPSHOTCLASS          VOLUMEGROUPSNAPSHOTCONTENT                              CREATIONTIME   AGE
my-group-snapshot-test                                   true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-fefe8683-c26c-47c6-95fe-a8a9762fc5ac   6d             6d
my-group-snapshot-test-1                                 true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-470adc84-d6b8-455d-a878-ed5b3665534f   6d             6d
my-group-snapshot-test-2                                 true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-e7684dc3-1487-4d86-bb7f-0ea7176e4c17   6d             6d
my-group-snapshot-test-3                                 true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-28ae0856-7331-41e1-87b8-172db0d27dbf   5d23h          5d23h
my-group-snapshot-test-4                                 true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-26e678c0-a320-460f-9bcd-0646bff04748   5d23h          5d23h
my-group-snapshot-test-55                                true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-84843f27-9934-4430-8e71-e7312538115d   5d23h          5d23h
my-group-snapshot-test-55333                             true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-867f07c1-6cc5-4dd1-935f-565fac55f322   5d23h          5d23h
my-group-snapshot-test-55334443                          true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-0f3fde21-0eea-438c-902c-44d4cb7a1f2e   5d22h          5d22h
my-group-snapshot-test-553344sfsd43                      true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-90f2744c-e2ad-47f2-a80a-3d26caa3dd41   5d22h          5d22h
my-group-snapshot-test-553344sfssdgfsdgfd43              true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-36a8f3a2-2c91-4db4-9d11-5a170a210faa   5d22h          5d22h
my-group-snapshot-test-553344sfssdgfsdgsdgffd43          true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-7c3e1c89-67d0-402b-b90d-c5778904546d   5d21h          5d21h
my-group-snapshot-test-553344sfssdgfsdgsdgsdgsdgfffd43   true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-7f53a378-cec7-41ce-b96b-2b983dabb92f   5d21h          5d21h
[🎩︎]mrajanna@li-2cfbef4c-22d9-11b2-a85c-a3e4a93c405f cephfs $]kubectl get volumegroupsnapshotcontent
NAME                                                    READYTOUSE   DELETIONPOLICY   DRIVER                          VOLUMEGROUPSNAPSHOTCLASS          VOLUMEGROUPSNAPSHOTNAMESPACE   VOLUMEGROUPSNAPSHOT                                      AGE
groupsnapcontent-0f3fde21-0eea-438c-902c-44d4cb7a1f2e   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-55334443                          5d22h
groupsnapcontent-26e678c0-a320-460f-9bcd-0646bff04748   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-4                                 5d23h
groupsnapcontent-28ae0856-7331-41e1-87b8-172db0d27dbf   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-3                                 5d23h
groupsnapcontent-36a8f3a2-2c91-4db4-9d11-5a170a210faa   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-553344sfssdgfsdgfd43              5d22h
groupsnapcontent-470adc84-d6b8-455d-a878-ed5b3665534f   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-1                                 6d
groupsnapcontent-7c3e1c89-67d0-402b-b90d-c5778904546d   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-553344sfssdgfsdgsdgffd43          5d21h
groupsnapcontent-7f53a378-cec7-41ce-b96b-2b983dabb92f   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-553344sfssdgfsdgsdgsdgsdgfffd43   5d21h
groupsnapcontent-84843f27-9934-4430-8e71-e7312538115d   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-55                                5d23h
groupsnapcontent-867f07c1-6cc5-4dd1-935f-565fac55f322   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-55333                             5d23h
groupsnapcontent-90f2744c-e2ad-47f2-a80a-3d26caa3dd41   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-553344sfsd43                      5d22h
groupsnapcontent-e7684dc3-1487-4d86-bb7f-0ea7176e4c17   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test-2                                 6d
groupsnapcontent-fefe8683-c26c-47c6-95fe-a8a9762fc5ac   true         Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-groupsnapclass   default                        my-group-snapshot-test                                   6d
[🎩︎]mrajanna@li-2cfbef4c-22d9-11b2-a85c-a3e4a93c405f cephfs $]kubectl get volumesnapshot
NAME                                                                                            READYTOUSE   SOURCEPVC    SOURCESNAPSHOTCONTENT                                                                              RESTORESIZE   SNAPSHOTCLASS                SNAPSHOTCONTENT                                                                                    CREATIONTIME   AGE
cephfs-pvc-snapshot                                                                             true         cephfs-pvc                                                                                                      1Gi           csi-cephfsplugin-snapclass   snapcontent-151352ad-c56c-4337-b9c2-b94f17ecb9f1                                                   6d             6d
snapshot-0bfc4b0c37dc5a75cff1168469b77b2233bb3378f872c7edd155d593b9f62471-2024-02-08-1.46.27    true                      snapcontent-0bfc4b0c37dc5a75cff1168469b77b2233bb3378f872c7edd155d593b9f62471-2024-02-08-1.46.27    0                                          snapcontent-0bfc4b0c37dc5a75cff1168469b77b2233bb3378f872c7edd155d593b9f62471-2024-02-08-1.46.27    5d23h          5d23h
snapshot-3293880fa07dcfea269a806651cd0cb7f0eec8bd43539b7076b62a3e8506a06e-2024-02-08-2.26.24    true                      snapcontent-3293880fa07dcfea269a806651cd0cb7f0eec8bd43539b7076b62a3e8506a06e-2024-02-08-2.26.24    0                                          snapcontent-3293880fa07dcfea269a806651cd0cb7f0eec8bd43539b7076b62a3e8506a06e-2024-02-08-2.26.24    5d23h          5d23h
snapshot-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.45     true                      snapcontent-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.45     1Gi                                        snapcontent-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.45     5d21h          5d21h
snapshot-4710ef52ed63bae2704e9ac756f92f5e8b3bd0895088d9b608bb3cf959104aef-2024-02-08-2.19.0     true                      snapcontent-4710ef52ed63bae2704e9ac756f92f5e8b3bd0895088d9b608bb3cf959104aef-2024-02-08-2.19.0     1Gi                                        snapcontent-4710ef52ed63bae2704e9ac756f92f5e8b3bd0895088d9b608bb3cf959104aef-2024-02-08-2.19.0     5d23h          5d23h
snapshot-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.3     true                      snapcontent-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.3     1Gi                                        snapcontent-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.3     5d22h          5d22h
snapshot-71fe32558053845f6e7bc446be564e7484415b9a847a86a174a45b47d301f9c9-2024-02-08-1.9.55     true                      snapcontent-71fe32558053845f6e7bc446be564e7484415b9a847a86a174a45b47d301f9c9-2024-02-08-1.9.55     0                                          snapcontent-71fe32558053845f6e7bc446be564e7484415b9a847a86a174a45b47d301f9c9-2024-02-08-1.9.55     6d             6d
snapshot-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.27   true                      snapcontent-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.27                                              snapcontent-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.27   26h            26h
snapshot-9d5577b49af06eac679eccdce34190246b3f75b29ad07eb7fc60fc5ef49ca0c3-2024-02-08-2.50.16    true                      snapcontent-9d5577b49af06eac679eccdce34190246b3f75b29ad07eb7fc60fc5ef49ca0c3-2024-02-08-2.50.16    0                                          snapcontent-9d5577b49af06eac679eccdce34190246b3f75b29ad07eb7fc60fc5ef49ca0c3-2024-02-08-2.50.16    5d22h          5d22h
snapshot-b5e389c7ef4f0ee830df7e932f28a61ca53b1330bbce37dbdcad0078da5b9883-2024-02-08-12.45.41   true                      snapcontent-b5e389c7ef4f0ee830df7e932f28a61ca53b1330bbce37dbdcad0078da5b9883-2024-02-08-12.45.41   0                                          snapcontent-b5e389c7ef4f0ee830df7e932f28a61ca53b1330bbce37dbdcad0078da5b9883-2024-02-08-12.45.41   6d             6d
snapshot-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.19    true                      snapcontent-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.19    1Gi                                        snapcontent-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.19    5d21h          5d21h
snapshot-ead7769081146a808774c39a00baafe2e948942544eb178247d181410c2039b6-2024-02-08-2.41.44    true                      snapcontent-ead7769081146a808774c39a00baafe2e948942544eb178247d181410c2039b6-2024-02-08-2.41.44    0                                          snapcontent-ead7769081146a808774c39a00baafe2e948942544eb178247d181410c2039b6-2024-02-08-2.41.44    5d22h          5d22h
snapshot-f39d8b0823d43ab28d7005c4c356331d51bd2685891461157650af28732cb88b-2024-02-08-1.1.42     true                      snapcontent-f39d8b0823d43ab28d7005c4c356331d51bd2685891461157650af28732cb88b-2024-02-08-1.1.42     0                                          snapcontent-f39d8b0823d43ab28d7005c4c356331d51bd2685891461157650af28732cb88b-2024-02-08-1.1.42     6d             6d
snapshot-f4977bcd648f6c1e419d2c343d506c0f55cc17c76ee9bdba0367c612a2f56a8d-2024-02-08-2.8.44     true                      snapcontent-f4977bcd648f6c1e419d2c343d506c0f55cc17c76ee9bdba0367c612a2f56a8d-2024-02-08-2.8.44     0                                          snapcontent-f4977bcd648f6c1e419d2c343d506c0f55cc17c76ee9bdba0367c612a2f56a8d-2024-02-08-2.8.44     5d23h          5d23h
[🎩︎]mrajanna@li-2cfbef4c-22d9-11b2-a85c-a3e4a93c405f cephfs $]kubectl get volumesnapshotcontent
NAME                                                                                               READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                          VOLUMESNAPSHOTCLASS          VOLUMESNAPSHOT                                                                                  VOLUMESNAPSHOTNAMESPACE   AGE
snapcontent-0bfc4b0c37dc5a75cff1168469b77b2233bb3378f872c7edd155d593b9f62471-2024-02-08-1.46.27    true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-0bfc4b0c37dc5a75cff1168469b77b2233bb3378f872c7edd155d593b9f62471-2024-02-08-1.46.27    default                   5d23h
snapcontent-151352ad-c56c-4337-b9c2-b94f17ecb9f1                                                   true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-snapclass   cephfs-pvc-snapshot                                                                             default                   6d
snapcontent-3293880fa07dcfea269a806651cd0cb7f0eec8bd43539b7076b62a3e8506a06e-2024-02-08-2.26.24    true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-3293880fa07dcfea269a806651cd0cb7f0eec8bd43539b7076b62a3e8506a06e-2024-02-08-2.26.24    default                   5d23h
snapcontent-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.43     true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.43     default                   5d21h
snapcontent-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.45     true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-33b917f2011cdb7d4c29624fbb3293e3be1ac48384c143f99ea6813008e5dfac-2024-02-08-4.9.45     default                   5d21h
snapcontent-4710ef52ed63bae2704e9ac756f92f5e8b3bd0895088d9b608bb3cf959104aef-2024-02-08-2.19.0     true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-4710ef52ed63bae2704e9ac756f92f5e8b3bd0895088d9b608bb3cf959104aef-2024-02-08-2.19.0     default                   5d23h
snapcontent-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.0     true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.0     default                   5d22h
snapcontent-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.3     true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-68e4bb33b198f852323ff3ec807bad20308e3dc48ea3c6bcde132f0166d4cbc0-2024-02-08-2.32.3     default                   5d22h
snapcontent-71fe32558053845f6e7bc446be564e7484415b9a847a86a174a45b47d301f9c9-2024-02-08-1.9.55     true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-71fe32558053845f6e7bc446be564e7484415b9a847a86a174a45b47d301f9c9-2024-02-08-1.9.55     default                   6d
snapcontent-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.25   true                       Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.25   default                   26h
snapcontent-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.27   true                       Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-83c197640e2f90695cfbf6164299fcd3da66713a8eef8d1d8fbdba6dfa089de8-2024-02-13-11.26.27   default                   26h
snapcontent-9d5577b49af06eac679eccdce34190246b3f75b29ad07eb7fc60fc5ef49ca0c3-2024-02-08-2.50.16    true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-9d5577b49af06eac679eccdce34190246b3f75b29ad07eb7fc60fc5ef49ca0c3-2024-02-08-2.50.16    default                   5d22h
snapcontent-b5e389c7ef4f0ee830df7e932f28a61ca53b1330bbce37dbdcad0078da5b9883-2024-02-08-12.45.41   true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-b5e389c7ef4f0ee830df7e932f28a61ca53b1330bbce37dbdcad0078da5b9883-2024-02-08-12.45.41   default                   6d
snapcontent-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.18    true                       Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.18    default                   5d21h
snapcontent-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.19    true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-bfe869219e3f8f6666277b1b83c35947174b0b1a13ec0e42195edde54e2c3e97-2024-02-08-4.14.19    default                   5d21h
snapcontent-ead7769081146a808774c39a00baafe2e948942544eb178247d181410c2039b6-2024-02-08-2.41.44    true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-ead7769081146a808774c39a00baafe2e948942544eb178247d181410c2039b6-2024-02-08-2.41.44    default                   5d22h
snapcontent-f39d8b0823d43ab28d7005c4c356331d51bd2685891461157650af28732cb88b-2024-02-08-1.1.42     true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-f39d8b0823d43ab28d7005c4c356331d51bd2685891461157650af28732cb88b-2024-02-08-1.1.42     default                   6d
snapcontent-f4977bcd648f6c1e419d2c343d506c0f55cc17c76ee9bdba0367c612a2f56a8d-2024-02-08-2.8.44     true         0             Delete           rook-ceph.cephfs.csi.ceph.com                                snapshot-f4977bcd648f6c1e419d2c343d506c0f55cc17c76ee9bdba0367c612a2f56a8d-2024-02-08-2.8.44     default                   5d23h
```

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,63 @@
+
+// validateVolumeGroupSnapshotRequest validates the request for creating a group
+// snapshot of volumes.
+func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
```
can you add a unit-test for this please?

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	if len(req.GetSourceVolumeIds()) == 0 {
+		return status.Error(codes.InvalidArgument, "source volume id cannot be empty")
```
id -> IDs, plural

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,193 @@
+		state = Quiescing
+	case "QUIESCED":
+		state = Quiesced
+	}
```
should there not be any validation, maybe with a `default` case?

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,193 @@
+	return volName
+}
+
+func (svq *subVolumeQuiesce) SubvolumeQuiesce(
```
`SubvolumeQuiesce` inconsisten use of `subVolume`, `SubVolume` and `Subvolume`

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,193 @@
+	fsa                   *admin.FSAdmin
+}
+
+func NewSubVolumeQuiesce(
```
a comment about the parameters would be nice, especially the `mapping` one

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,193 @@
+	return svq.volumes
+}
+
+func (svq *subVolumeQuiesce) getSubvolumeNamesString() []string {
```
Please explain in a comment what this function does. What is the original contents look like, and how are the `volName` (plural?)` constructed?

Madhu-1 replied:
updated function name to getMembers(), `ceph fs quiesce` CLI calls this as members and added a comment for it with example

nixpanic commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,320 @@
+	"github.com/container-storage-interface/spec/lib/go/csi"
+)
+
+type VolumeGroupOptions struct {
```
would it make sense to inherit VolumeOptions, or create a new VolumeOptionsBase struct with shared functions?

Madhu-1 replied:
inherited for now.

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+			// progress if we execute the quiesce again until the QUIESCED is
+			// reached or any other error state is reached
+			log.ErrorLog(ctx, "failed to quiesce subvolume: %v", err)
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
```
Undoing the reservation here does not look like the right place. Ideally it should be done in the same function where the reservation was created (maybe in a cleanup helper function).

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+		}
+		state := core.GetQuiesceState(data)
+		if state != core.Released {
+			return nil, fmt.Errorf("quiesce operation is in %s state", state)
```
returning a non-gRPC error here

nixpanic replied:
In general my preference goes to only returning gRPC errors in CSI procedures. Helper functions should just return 'normal' errors.

Madhu-1 replied:
fixed non-grpc error returning, the helper functions inside `groupcontrollerserver.go` are the only one returning grpc error as little much as possible to avoid creating and passing/extracting own errors to identify what is the grpc error codes to pass.

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+	requestName string,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+) error {
+	releasing := false
```
Naming `releasing` seems a little confusing, isn't it the opposite? Maybe use `inProgress` or something?

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+	err = store.VolumeGroupJournal.AddVolumeSnapshotMapping(ctx,
+		vgo.MetadataPool,
+		vgs.ReservedID,
+		req.SourceVolumeId,
```
all `req` usage should ideally use `.Get..()` functions for improved error handling/parsing in the gRPC layer

nixpanic replied:
all `req` usage should ideally use `.Get..()` functions for improved error handling/parsing in the gRPC layer.

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+	return ""
+}
+
+func getFsNamesAndSubVolumeFromVolumeIDs(ctx context.Context,
```
long functions like this benefit a lot from comments explaining what/how the do it.

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+	return fsk, nil
+}
+
+func matchesSourceVolumeIDs(sourceVolumeIDs, volumeIDsInOMap []string) bool {
```
interesting that there is no standard function for this, consider using `Equal()` from `k8s.io/utils/strings/slices`?

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -735,3 +735,11 @@ func extractDeleteVolumeGroupError(err error) error {
 	return err
 }
+
+// GetVolumeGroupSnapshot return Unimplemented error as its not implemented.
```
This should already be inherited from the UnimplementedControllerServer struct that the CSI grpc libs provide.

Madhu-1 committed:
cephfs: unit test for validateVolumeGroupSnapshotRequest

Added unit test for
validateVolumeGroupSnapshotRequest API which
validates the input VolumeGroupSnapshotRequest
request

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@nixpanic Thanks for the review, i added all the reviews PTAL.

nixpanic commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,310 @@
+}
+
+// Connect a CephFS volume to the Ceph cluster.
+func (vo *VolumeGroupOptions) Connect(cr *util.Credentials) error {
```
how is this different from `VolumeOptions.Connect()`? Same for other functions.

Madhu-1 replied:
Yes this will be same as `VolumeOptions.Connect()` should have removed when we embedded `VolumeOptions` to volumeGroupOptions`. removed both connect and Destroy for now.

nixpanic commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,310 @@
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	cr *util.Credentials,
+) (*VolumeGroupOptions, error) {
```
Would it make sense to call `NewVolumeOptions()` and type-cast the result to `VolumeGroupOptions`?

Madhu-1 replied:
NewVolumeOptions expects input as CreateVolumeRequest and also it doesn't have many checks that are not required for the group operations and there are few strict checks when it comes to VolumeOptions but are not required for groupOptions.

nixpanic replied:
Can the shared checks be moved into helper functions, so that the duplication gets reduced at lease a little?

Madhu-1 replied:
sure that can be done, updated it now.

nixpanic commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,286 @@
+
+	opts.RequestName = req.GetName()
+
+	err = opts.Connect(cr)
```
there is no matching `Destroy()` in case an error occurs later on, this may leak a connection  to the Ceph cluster.

Madhu-1 replied:
Yes correct, Destroy was missing in err case, fixed now.

Madhu-1 committed:
cephfs: add helper function to getVolumeOptions

added helper function to extract basic
details from the parameters related to volume
options.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
build: add ceph_preview to GO_TAGS_LIST

added required ceph_preview tag to the
GO_TAGS_LIST in Makefile which is
required for FSQuiesce API.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Rakshith-R commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,216 @@
+
+func (fq *fsQuiesce) FSQuiesce(
+	ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Timeout:    180,
+		AwaitFor:   0,
+		Expiration: 180,
+	}
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to quiesce filesystem %s", err)
+
+		return nil, err
+	}
+	qInfo := resp.Sets[reserveName]
+
+	return &qInfo, nil
+}
+
+func (fq *fsQuiesce) FSQuiesceWithExpireTimeout(ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Timeout:    180,
+		AwaitFor:   0,
+		Expiration: 180,
+	}
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to quiesce filesystem with expire timeout %s", err)
+
+		return nil, err
+	}
+
+	qInfo := resp.Sets[reserveName]
+
+	return &qInfo, nil
+}
```
Both these functions are exactly the same ?

Madhu-1 replied:
Yes for now it looks the same but the functionality is different, to keep it simple i have created two helpers without if/else statements

Rakshith-R commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,216 @@
+	return &qInfo, nil
+}
+
+func (fq *fsQuiesce) ResetFSQuiesce(ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Reset:      true,
+		AwaitFor:   0,
+		Timeout:    180,
+		Expiration: 180,
+	}
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if err != nil {
```
Please add some comments regarding what reset does. 
Do we need timeout and expiration here too?

Madhu-1 replied:
Yes it takes the timeout value and added comment what it does

Rakshith-R commented on internal/cephfs/errors/errors.go:
```diff
@@ -58,6 +58,9 @@ var (
 	// ErrVolumeHasSnapshots is returned when a subvolume has snapshots.
 	ErrVolumeHasSnapshots = coreError.New("volume has snapshots")
+
+	// ErrInProgress is returned when quiesce operation is in progress.
+	ErrInProgress = coreError.New("quiesce operation in progress")
```
```suggestion
	// ErrQuiesceInProgress is returned when quiesce operation is in progress.
	ErrQuiesceInProgress = coreError.New("quiesce operation is in progress")
```

Rakshith-R commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,218 @@
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Timeout:    180,
+		AwaitFor:   0,
+		Expiration: 180,
```
Can we use Constants since same values are used at multiple places ?

Madhu-1 replied:
IMO we cannot reuse the same timeout for different function, that's the reason for not keeping it constant

Rakshith-R commented on internal/cephfs/store/fsjournal.go:
```diff
@@ -40,6 +40,10 @@ var (
 	SnapJournal *journal.Config
+
+	// VolumeGroupJournal is used to maintain RADOS based journals for CO
+	// generated.
+	VolumeGroupJournal journal.VolumeGroupJournal
```
```suggestion
	// generated request name to CephFS snapshot group attributes.
	VolumeGroupJournal journal.VolumeGroupJournal
```

Just ending with ` for CO generated.` seems incomplete.

Rakshith-R commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,285 @@
+		return nil, err
+	}
+
+	func() {
```
```suggestion
	defer func() {
```
defer is missing ?

Madhu-1 replied:
good catch 👍🏻

Rakshith-R reviewed (commented):
few more small nits. 

(just 2 more large files yet to be reviewed :sweat_smile: )

Rakshith-R commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,756 @@
+	inProgress, err := cs.queisceFileSystems(ctx, vgs, fsMap)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to quiesce filesystems: %v", err)
+		if err != nil && !errors.Is(err, cerrors.ErrQuiesceInProgress) {
```
```suggestion
		if !errors.Is(err, cerrors.ErrQuiesceInProgress) {
```

Rakshith-R commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,756 @@
+		log.ErrorLog(ctx, "failed to create snapshot and add to volume group journal: %v", err)
+
+		if !errors.Is(err, cerrors.ErrQuiesceInProgress) {
+			log.ErrorLog(ctx, "failed to create snapshot and add to volume group journal: %v", err)
```
```suggestion
```
same error is logged just above.

Rakshith-R commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,756 @@
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", uErr)
+			}
+
+			return nil, status.Error(codes.Internal, err.Error())
```
```suggestion
```
same return exists just below

Rakshith-R commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,756 @@
+
+			return inProgress, err
+		}
+		state := core.GetQuiesceState(data.State)
+		if state == core.Quiescing {
+			inProgress = true
+		} else if state != core.Quiesced {
+			return inProgress, fmt.Errorf("quiesce operation is in %s state", state)
+		}
```
Is this function supposed to error out if the state is not equal to ( either quiesced or quiescing ) ?

Madhu-1 replied:
Yes thats correct, the state should be either in `Quiescing` or `Quiesced` nothing else

nixpanic replied:
why not just write `else` in that case? The `if state != core.Quiesced` makes it a little more difficult to read and can cause confusion.

Madhu-1 replied:
else will also consider the `core.Quiesced` and return an error message, for that reason using else if. instead of else.

nixpanic replied:
ah, right, `core.Quiescing != core.Quiesced` :mag: 

A switch with the different states would be cleaner, or a simple `inProgress, err := state.IsInProgress()` function?

Rakshith-R commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,240 @@
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, []string{}, reserveName, opt)
+	log.DebugLog(ctx, "error and response", resp, err)
```
```suggestion
	log.DebugLog(ctx, "error and response", resp, err)
```

This line may need to be removed if it was added for debugging during coding or error message may need to be improved.

Madhu-1 replied:
it was added for testing, removed it.

Rakshith-R commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,756 @@
+		if state != core.Released {
+			inProgress = true
+
+			continue
```
This continue is not required ? 
or is some parts of code missing ?

Madhu-1 replied:
it can be removed not required

Rakshith-R commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,756 @@
+		state := core.GetQuiesceState(data.State)
+		if state == core.Quiescing {
+			inProgress = true
+
+			continue
+		}
+		if state != core.Quiesced {
```
```suggestion
		} else if state != core.Quiesced {
```

I think this makes it simpler. 
Let's go with this only if it makes sense to you too.

Madhu-1 committed:
rebase: update go-ceph to latest commit

updating go-ceph to latest commit to
pull the changes required for ceph fs
quiesce. This is also updating aws
sdk dependency.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@nixpanic @Rakshith-R This is ready for re-review and to approve. i have done scale testing of creating VSG for 50 PVC all seems to be working fine as expected.

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,237 @@
+	"github.com/ceph/go-ceph/cephfs/admin"
+)
+
+type QuiesceStates string
```
nit: should probably be called `QuiesceState`

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,237 @@
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if resp != nil {
```
We usually write code that returns early when an error happened and continues on success. Maybe consider changing the style here to what we do? Not sure if it is in the coding guidelines, it is not a blocker for me.

Madhu-1 replied:
we get both error and response, things are little easier based on the response, it becomes complicated to consider both the error and response to take some calls, i would prefer to keep it this way for now which rather simply the checks

nixpanic commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,285 @@
+	vgs.ReservedID = vi.ObjectUUID
+
+	if volOptions.Monitors, err = util.Mons(util.CsiConfigFile, vi.ClusterID); err != nil {
+		return volOptions, &vgs, fmt.Errorf(
```
Are the returned `volOptions` and `vgs` useful for the caller in case of an error? I prefer to return `nil` and not return incomplete/broken structs.

nixpanic commented on internal/cephfs/store/volumegroup.go:
```diff
@@ -0,0 +1,285 @@
+		return volOptions, &vgs, err
+	}
+
+	err = VolumeGroupJournal.Connect(volOptions.Monitors, fsutil.RadosNamespace, cr)
```
Is `VolumeGroupJournal` a global struct? What happens when 2 different requests are coming in at the same time, possibly even with different parameters like MONs? Can this global struct be connected multiple times without its `Destroy()` causing issues?

Madhu-1 replied:
Yes this can cause issues, i have updated to code to create new connections to avoid this problem

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,566 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+		GroupSnapshotId: vgs.VolumeGroupSnapshotID,
+	}
+	response.GroupSnapshot.ReadyToUse = true
+	response.GroupSnapshot.CreationTime = timestamppb.New(time.Now())
```
nit: These two values can be put inside the `&csi.VolumeGroupSnapshot{` initializer just like `GroupSnapshotId`. It would be cleaner to do that.

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,566 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+	response.GroupSnapshot = &csi.VolumeGroupSnapshot{
+		GroupSnapshotId: vgs.VolumeGroupSnapshotID,
+	}
+	response.GroupSnapshot.ReadyToUse = true
```
again, nit: why not include them into the `&csi.VolumeGroupSnapshot{` initializer?

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +72,566 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	cr *util.Credentials,
+	fsMap map[string]core.FSQuiesceClient,
+	secrets map[string]string) (
```
Wow, one of the previous functions has 6 parameters, and I considered mentioning that. This one has :eight:! It becomes rather unmanageable now... Maybe it makes sense to split the functionality of this function into smaller pieces to enhance the maintainability.

Madhu-1 replied:
Removed one parameter for now, Taking this as action items to refactor it, it looks like its not easy to refactor as the internal function needs more parameter

Madhu-1 replied:
Refactored now to remove 2 parameters, now both will have 6 but will try to reduce it (if possible)

Madhu-1 committed:
journal: fix connection problem with groupjournal

Same group jounral config need to be reused
for multiple connection where different monitors
and users are used, for that reason create a unique
connection each time.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Rakshith-R commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,237 @@
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if resp != nil {
+		qInfo := resp.Sets[reserveName]
+
+		return &qInfo, nil
+	}
+
+	log.ErrorLog(ctx, "failed to reset timeout for quiesce filesystem %s", err)
+
+	return nil, err
```
```suggestion
	if err != nil{
		log.ErrorLog(ctx, "failed to reset timeout for quiesce filesystem %s", err)

		return nil, err		
	}
	if resp == nil {
		return nil, fmt.Errorf("response is nil")
	}
	qInfo := resp.Sets[reserveName]

	return &qInfo, nil
```

This may lead to a panic in case both error and resp is nil ? 
I think its better to check both just in case ?

Madhu-1 replied:
IMHo I don't really see a case when both can be nil, I expect at least one to be set, if it's really required I can make that change

Rakshith-R reviewed (dismissed):
LGTM, Thanks !

nixpanic commented on internal/cephfs/core/quiesce.go:
```diff
@@ -0,0 +1,237 @@
+	return volName
+}
+
+func (fq *fsQuiesce) FSQuiesce(
```
What is the difference with `FSQuiesceWithExpireTimeout()`? If both functions are needed, please comment them so that it is clear when to use one of them, and not the other.

Madhu-1 replied:
added a comment when to use `FSQuiesceWithExpireTimeout`

nixpanic replied:
Thanks, but I still do not see a difference in the functions. Both seem to pass the same Timeout + Expiration?

Madhu-1 replied:
Yes for now it looks the same, based on the testing we might change the timeout and expiration for both, for easy reading/maintenance i have kept 2 separate functions for now.

nixpanic replied:
If this is a TODO, please add that as a comment, so that grepping for TODO in the sources finds it again.

Not sure if it is beneficial to keep two identical functions. Please consider removing the duplicate and commenting TODO in the code where a (non)timeout function may be preferable.

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,758 @@
+	secrets map[string]string,
+) error {
+	// get the omap from the snapshot and volume mapping
+	bgi, vgsi, err := store.NewVolumeGroupOptionsFromID(ctx, vgs.VolumeGroupSnapshotID, cr)
```
`NewVolumeGroupOptionsFromID()` returns a connected `VolumeGroupOptions`. A `defer bgi.Destroy()` seems to be missing.

Madhu-1 replied:
yes its a good catch, fixed it now

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,758 @@
+	}
+	defer cr.DeleteCredentials()
+
+	_, vgsi, err := store.NewVolumeGroupOptionsFromID(ctx, req.GroupSnapshotId, cr)
```
The `_` variable needs a `Destroy()` here too, I think?

Madhu-1 replied:
yes its a good catch, fixed it now

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -0,0 +1,760 @@
+	}
+	defer cr.DeleteCredentials()
+
+	vg, err := store.NewVolumeGroupOptions(ctx, req, cr)
```
`vg.Destroy()` seems to be missing here

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +73,585 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+	var err error
+	for k, v := range fm {
+		conn := &util.ClusterConnection{}
+		if err = conn.Connect(v.monitors, cr); err != nil {
```
this `conn.Connect()` does not seem to have a matching `conn.Destroy()`?

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -655,3 +655,107 @@ func (cs *ControllerServer) deleteSnapshotsAndUndoReservation(ctx context.Contex
+	vgo.Destroy()
+
+	volIds := vgsi.GetVolumeIDs()
+	fsMap, err := getFsNamesAndSubVolumeFromVolumeIDs(ctx, req.GetSecrets(), volIds, cr)
```
fsMap has connections, I think? They need to call `conn.Destroy()` somewhere.

Madhu-1 replied:
Yes it was missing again, fixed it now. :(

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +73,601 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer destroyFSConnections(fsMap)
```
It would be nice if `fsMap.Destroy()` can be used. Returning the `fsMap` with a defined type and not a generic `map[string]core.FSQuiesceClient` can probably be done. That can then be used to cleanup the code a little more, and reduce the number of helper functions, so that they are combined directly with a `fsMap` type.

A quick check that something like this works, a `map` should not be much different.

```go
package main

import (
        "fmt"
)

type object struct {
        name string
}

type objectList []*object

func (o *object) String() string {
        return "my name is " + o.name
}

func newObjects() objectList {
        return objectList{
                {name: "first"},
                {name: "second"},
                {name: "third"},
        }   
}

func (l objectList) String() string {
        s := ""

        for i, o := range l { 
                s += fmt.Sprintf("#%d: %s\n", i, o)
        }   

        return s
}

func main() {
        list := newObjects()

        fmt.Println(list)
}
```

nixpanic commented on internal/cephfs/groupcontrollerserver.go:
```diff
@@ -61,3 +73,601 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+
+	responses := make([]csi.CreateSnapshotResponse, 0)
+	for _, volID := range req.GetSourceVolumeIds() {
+		err = fsQuiesceWithExpireTimeout(ctx, vgo.RequestName, fsMap)
```
In lots of places where `fsMap` is passed, it might be nice to have functions like `fsMap.fsQuiesceWithExpireTimeout(...)`

nixpanic reviewed (approved):
I can not spot any major issues anymore, thanks for the corrections!

The left-over comments are things that will improve maintainability, and can be sent as follow-up PRs.

Madhu-1 commented:
> I can not spot any major issues anymore, thanks for the corrections!
> 
> The left-over comments are things that will improve maintainability, and can be sent as follow-up PRs.

Sure will address as follow-up's, Thanks @nixpanic

Madhu-1 commented:
@mergifyio queue
